src/sentry/tasks/assemble.py

```diff
@@ -181,293 +257,228 @@ class AssembleArtifactsError(Exception):
     pass
 
 
-def _simple_update(
-    release_file: ReleaseFile, new_file: File, new_archive: ReleaseArchive, additional_fields: dict
-) -> bool:
-    """Update function used in _upsert_release_file"""
-    old_file = release_file.file
-    release_file.update(file=new_file, **additional_fields)
-    old_file.delete()
+class PostAssembler(ABC):
+    def __init__(self, assemble_result: AssembleResult):
+        self.assemble_result = assemble_result
+        self._validate_assemble_result()
 
-    return True
+    def _validate_assemble_result(self):
+        if self.assemble_result.bundle is None or self.assemble_result.temp_file is None:
+            raise AssembleArtifactsError("the assemble result has invalid bundle or temporary file")
```

Comment:
I would rather filter this case before you pass an invalid `AssembleResult` into the constructor.

You are also a bit inconsistent with error handling. You are explicitly deleting the created `file` in both of the `_validate_bundle` methods, but you are not doing so when the `post_assemble` fails.

Will `post_assemble` ever be retried, or what would happen with the dangling `File` in a failure case? IMO we should guard the whole block around `prepare_post_assembler` and clean up the `File` in that case as well.

Reply:
I agree, will take care of everything!